### PR TITLE
Faster zip hashing, take 2

### PR DESCRIPTION
**Pull request by oehme**
This reapplies #9344, which was reverted because of a memory leak
in java.util.ZipFile. The leak was fixed in recent Java versions, so the new
version of this code only uses that class on new Java versions.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+
+public class ZipHasher implements RegularFileHasher, ConfigurableNormalizer {
+
+    private static final Set<String> KNOWN_ZIP_EXTENSIONS = ImmutableSet.of(".zip", ".jar", ".war", ".rar", ".ear", ".apk", ".aar");
```
Comment (paplorinc) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java]: why is `.rar` a zip file?

Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java]: That's a Java EE `Resource Adapter Archive `

Comment (paplorinc) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java]: got it, thanks.

**Comment by oehme** [comment]
I've changed the approach again, this time only using ZipFile on the later Java versions, where we know it works. This did the trick on my test run on CI.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileZipInput.java [hunk 2]
```diff
@@ -29,10 +32,29 @@
...
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    /**
+     * {@link ZipFile} is more efficient, but causes memory leaks on older Java versions, so we only use it on more recent ones.
+     */
+    private static boolean isZipFileSafeToUse() {
+        return JavaVersion.current().isJava11Compatible();
```
Comment (paplorinc) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileZipInput.java]: is there a bug report that we can attach here?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileZipInput.java [hunk 3]
```diff
@@ -29,10 +32,29 @@
 
 class FileZipInput implements ZipInput {
 
+    public static ZipInput create(File file) {
```
Comment (paplorinc) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileZipInput.java]: it's a bit weird that we have a factory that may or may not return itself.
Would it make sense to externalize the factory?

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileZipInput.java (hunk 2 above)
Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileZipInput.java]: Unfortunately not, we only know it works in recent Java versions

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbiExtractingClasspathResourceHasher.java [hunk 4]
```diff
@@ -70,8 +70,13 @@ public HashCode hash(ZipEntry zipEntry) throws IOException {
         if (!isClassFile(zipEntry.getName())) {
             return null;
         }
-        byte[] content = new byte[zipEntry.size()];
-        ByteStreams.readFully(zipEntry.getInputStream(), content);
+        byte[] content;
+        if (zipEntry.size() >= 0) {
```
Comment (paplorinc) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbiExtractingClasspathResourceHasher.java]: that's a possibility? I.e. they're not storing the size? How did you find out?

Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbiExtractingClasspathResourceHasher.java]: You don't get the size when you have a ZipInputStream. The size of each entry is only listed at the end of a zip file.

**Comment by CristianGM** [comment]
@lptr why is it reverted?

**Pull request merged**